src/api/options.c

```diff
@@ -628,6 +636,56 @@ nlopt_result NLOPT_STDCALL nlopt_get_xtol_abs(const nlopt_opt opt, double *xtol_
     return NLOPT_SUCCESS;
 }
 
+nlopt_result NLOPT_STDCALL nlopt_set_x_weights(nlopt_opt opt, const double *x_weights)
+{
+    if (opt) {
+        nlopt_unset_errmsg(opt);
+        if (!opt->x_weights) {
+            opt->x_weights = (double *) calloc(opt->n, sizeof(double));
+            if (!opt->x_weights) {
+                return NLOPT_OUT_OF_MEMORY;
+            }
+        }
+        memcpy(opt->x_weights, x_weights, opt->n * sizeof(double));
```

Comment:
It should probably check that all of the weights are > 0 or return an INVALID_ARGS